text/1028-cli-hooks.rst

```diff
@@ -122,23 +125,105 @@ means that the hooks associated with applying schema changes must be executed
 for these commands as well.
 
 
+Watch Configuration
+===================
+
+Sometimes in a Gel project there's a need to respond to some file changes: run
+a migration due to schema change or run code generators due to query file
+change. The CLI already supports watching the schema for changes and applying
+them, but we can generalize this mechanism and make it more flexible, much
+like the hooks that respond to CLI commands.
+
+The idea is to have a mechanism for specifying what changes you want to watch
+and a script that gets triggered by those changes. Unlike hooks, there are no
+"before" and "after" triggers here. Only one kind of response is possible:
+trigger when a change to the watched entity is detected.
+
+The watch configuration will be described in the ``gel.toml`` project manifest
+file in the ``[watch]`` table. We will introduce a ``[watch.files]`` sub-table
+for watching file system changes and triggering scripts. Potentially, we will
+also later introduce a ``[watch.gel-config]`` sub-table for monitoring changes
+to the various database config values and responding to them.
+
+The general structure of watch tables is going to use the key (potentially
+quoted) to specify *what is being watched* and the corresponding value to
+specify the script that will be triggered by changes.
+
+Since the triggered scripts will be executed in response to changes, we cannot
+meaningfully display the output of the scripts to the user. We may want to
+setup an optional logfile to record the script output. This can be specified
+in the general ``[watch]`` section as ``logfile="<path-to-logfile>"``.
+
+We may want to setup a bit of a delay before triggering the scripts so as to
+allow a sequence of changes to happen before the response script is triggered.
+This is mostly to reduce unnecessary multiple triggers for the same watched
+entity.
```

Comment:
Yeah, I was wondering a bit about overlapping globs like `**/*.edgeql` and `dbschema/*.edgeql` for instance: We don't necessarily want to try to sequence the scripts in a way that they affect each other, so blind parallel execution seems reasonable. Users will need to take care that their scripts do not cause conflicts with each other.

Reply:
So there are two problems:

- IDEs or other tools modifying/moving multiple files in one after another. The RFC says (and I agree) that all the events should trigger the watch script only once. To make this happen, we'll use this: https://docs.rs/notify-debouncer-mini/latest/notify_debouncer_mini/

- scripts modifying more files, which could recurse into the watch script again. This would be partially solved by a smart debouncer, but not entirely, since the watch script could first do some slow queries (say a second) and only then write the files. So I think it would be better to allow recursion, as you suggest @scotttrinh